### DISCUSSION TITLE
Can I use ListMigrations to show a list of Migrations on an MVC Page?

### DISCUSSION THREAD

**Discussion by asifshakir** [fluentmigrator/fluentmigrator#1439] [Q&A] [answered]
Hello,

Can I use ListMigrations to show a list of available migrations to the user on an ASP.NET Core MVC Website (using .NET Core 5.0)?

            var serviceProvider = CreateServices();
            using (var scope = serviceProvider.CreateScope())
            {
                var runner = serviceProvider.GetRequiredService<IMigrationRunner>();
                runner.ListMigrations();
            }

I believe I would need some way to fetch the output of ListMigrations and show pass it to the view.

**Comment by jzabroski** [comment, marked as answer]
@asifshakir MigrationRunner has a public constructor that takes an ILogger<MigrationRunner> instance:

https://github.com/fluentmigrator/fluentmigrator/blob/7033cdf353dcdcae721093a6d2cf646a2823da18/src/FluentMigrator.Runner/MigrationRunner.cs#L202

I could envision a "logger" component that has a public property which is the data structure for your ASP.NET MVC view.  

However, a cleaner solution would probably be to modularize MigrationRunner.  I would start with just what you need modularized, e.g., 

https://github.com/fluentmigrator/fluentmigrator/blob/7033cdf353dcdcae721093a6d2cf646a2823da18/src/FluentMigrator.Runner/MigrationRunner.cs#L912-L937

Something like the following should be a self-contained Service:

```c#
// Assume MigrationWithStatus is a DTO.
// Therefore, MigrationWithStatus  does not contain any methods or transitive references to objects that contain methods.
public ICollection<MigrationWithStatus> GetMigrationsWithStatus()
{
            var currentVersionInfo = VersionLoader.VersionInfo;
            var currentVersion = currentVersionInfo.Latest();

            var migrationsWithStatus = MigrationLoader
                        .LoadMigrations()
                        .Select(migration => {
                                    var migrationName = migration.Value.GetName();
                                    var status = GetStatus(migration , currentVersion);
                                    var statusStrings = GetStatusStrings(status);
                                    var statusString = string.Join(",", statusStrings);
                                    return
                                                new MigrationWithStatus() {
                                                            Name = migrationName,
                                                            Status = status,
                                                            StatusDetails = GetStatusStrings(status),
                                                            IsCurrent = (status & MigrationStatus.AppliedMask) == MigrationStatus.Current,
                                                            IsBreaking = (status & MigrationStatus.Breaking) == MigrationStatus.Breaking,
                                                            Message = $"{migrationName}{(string.IsNullOrEmpty(statusString) ? string.Empty : $" ({statusString})")}",
                                                };
                        })
                        .ToList();
            return migrationsWithStatus;
}
```

You would need to basically take my proof-of-concept above and,

1. Flesh out `MigrationWIthStatus` DTO - I recommend adding a get-only property that applies ViewModel logic such as `public bool IsEmphasized => IsCurrent || IsBreaking;`
2. Create a Service
3. Construct the Service inside the existing MigrationRunner constructor, for now, so that it is not a public breaking API change
4. Update the MigrationRunner.ListMigrations method to use the Service and write the appropriate Log messages
5. You can then add to the DI namespace a helper for also creating an IMyNewService - open to suggestions on what to call this.
6. Write tests for the above so that we have code coverage.
